sopel/plugins/rules.py

```diff
@@ -1301,3 +1358,140 @@ def parse(self, text):
+        ``match`` arguments will be the same when the rule executes.
+
+        This behavior makes the ``match`` parameter obsolete.
+
+    """
+    @classmethod
+    def from_callable(cls, settings, handler):
+        url_regexes = getattr(handler, 'url_regex', []) or []
+        if not url_regexes:
+            raise RuntimeError(
+                'Invalid URL callback: %s' % handler)
+
+        kwargs = cls.kwargs_from_callable(handler)
+
+        @functools.wraps(handler)
+        def execute_handler(bot, trigger):
+            return handler(bot, trigger, match=trigger)
+
+        kwargs.update({
+            'handler': execute_handler,
+            'schemes': settings.core.auto_url_schemes,
+        })
+
+        return cls(url_regexes, **kwargs)
+
+    @classmethod
+    def from_callable_lazy(cls, settings, handler):
+        """Instantiate a rule object from a handler with lazy-loaded regexes.
+
+        :param settings: Sopel's settings
+        :type settings: :class:`sopel.config.Config`
+        :param callable handler: a function-based rule handler with a
+                                 lazy-loader for the regexes
+        :return: an instance of this class created from the ``handler``
+        :rtype: :class:`AbstractRule`
+
+        Similar to the :meth:`from_callable` classmethod, it requires a rule
+        handlers decorated with :mod:`sopel.plugin`'s decorators.
+
+        Unlike the :meth:`from_callable` classmethod, the regexes are not
+        already attached to the handler: its loader function will be used to
+        get the rule's regexes. See the :func:`sopel.plugin.url_lazy` decorator
+        for more information about the handler and its loader signatures.
+        """
+        url_lazy_loader = getattr(handler, 'url_lazy_loader', [])
+        if not url_lazy_loader:
+            raise RuntimeError(
+                'Invalid lazy loader URL callback: %s' % handler)
+
+        regexes = tuple(url_lazy_loader(settings))
+        kwargs = cls.kwargs_from_callable(handler)
+        kwargs.update({
+            'handler': handler,
+            'schemes': settings.core.auto_url_schemes,
+        })
+
+        return cls(regexes, **kwargs)
+
+    def __init__(self,
+                 regexes,
+                 schemes=None,
+                 **kwargs):
+        super(URLCallback, self).__init__(regexes, **kwargs)
+        # prevent mutability of registered schemes
+        self._schemes = tuple(schemes or URL_DEFAULT_SCHEMES)
+
+    def match(self, bot, pretrigger):
+        """Match URL(s) in a pretrigger according to the rule.
+
+        :param bot: Sopel instance
+        :type bot: :class:`sopel.bot.Sopel`
+        :param pretrigger: Line to match
+        :type pretrigger: :class:`sopel.trigger.PreTrigger`
+
+        This method looks for :attr:`URLs in the IRC line
+        <sopel.trigger.PreTrigger.urls>`, and for each it yields
+        :func:`match objects <re.match>` using its regexes.
```

Comment:
See related note in `sopel/plugin.py` for example/etc.

```suggestion
        :ref:`match objects <match-objects>` using its regexes.
```